### PR TITLE
feat: Mark and punishment system

### PR DESCRIPTION
**Pull request by NjaalSoerland**
Closes: #177 

## Checklist

- [x] I have written tests
- [ ] I have provided documentation

## Changelog

- Added schemas for marks and personal marks
- Created a repository and service for each of them
- Implemented functionality for creating marks, and adding and removing them from users
- Added functions for mark expiration calculation

## How to test

yarn test

### CODE REVIEW

**Comment by kharann** [comment]
DATE FNS???

**Comment by NjaalSoerland** [comment]
> DATE FNS???

Wip ;)

**Review by kharann — CHANGES_REQUESTED** [review]
Please convert the database models from prisma to the format with kysely.

Read #232 for the reasoning behind it. You can see in #235 how you create migrations etc.

You need to:
- rebase
- Convert schema to migrations in `ow-db`
- Convert repositories and test to use kysely
- Write database types in `ow-db`
-

File: packages/types/src/mark.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  givenAt: z.date(),
+  updatedAt: z.date(),
+  category: z.string(),
+  details: z.string(),
+  givenTo: z.string().array(),
+  duration: z.number(),
+})
+
+export type Mark = z.infer<typeof MarkSchema>
+export type InsertMark = Omit<Mark, "id">
```
Comment (kharann) [inline packages/types/src/mark.ts:15]: Call it MarkWrite

File: packages/types/src/personal-mark.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { z } from "zod"
+
+export const personalMarkSchema = z.object({
+  markId: z.string(),
+  userId: z.string(),
+})
+
+export type PersonalMark = z.infer<typeof personalMarkSchema>
+export type InsertPersonalMark = PersonalMark
```
Comment (kharann) [inline packages/types/src/personal-mark.ts:9]: Rename to PersonalMarkWrite

**Review by kharann — CHANGES_REQUESTED** [review]

File: packages/api/src/modules/marks/mark-service.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+          personalMarkRepository.addPersonalMarkToUser(userId, markId),
+        ])
+        return markInfo
+      }
+      throw new Error(`User with ID:${userId} already has mark with ID:${markId}`)
+    },
+    deleteMark: async (id: string) => {
+      const mark = await service.getMark(id)
+      if (!mark) throw new NotFoundError(`Mark with ID:${id} not found`)
+      mark.givenTo.forEach(async (userId) => {
```
Comment (kharann) [inline packages/api/src/modules/marks/mark-service.ts:68]: You could set the Mark to `ON DELETE CASCADE`, which will automatically delete all related marks when the given mark is deleted

**Review by kharann — APPROVED** [review]

**Pull request merged**